File: elementExtensions/pl-drawing/resistor/resistor.py
import prairielearn as pl
import pygraphviz
import numpy as np
from os import path


# Import class definitions and default values from the drawing element
defaults = pl.load_host_script('defaults.py')
elements = pl.load_host_script('elements.py')

# Elements
class Resistor(elements.BaseElement):
    def generate(el, data):
        image = 'resistor-' + str(pl.get_integer_attrib(el, 'ohm', 1000)) + '.svg'
        return {
            'left': pl.get_float_attrib(el, 'x', 100),
            'top': pl.get_float_attrib(el, 'y', 100),
            'angle': pl.get_float_attrib(el, 'angle', 0),
            'ohm': pl.get_integer_attrib(el, 'ohm', 1000),
            'image_url': path.join(data['clientFilesUrl'], image)
        }

    def get_attributes():
        return ['x', 'y', 'ohm', 'angle']


elements = {}
elements['pl-resistor'] = Resistor

File: elementExtensions/pl-drawing/resistor/resistor.js
class Resistor extends PLDrawingBaseElement {
  static generate(canvas, options, submittedAnswer) {
    fabric.Image.fromURL(options.image_url, (obj) => {
      /* Generate a unique ID for this object if it doesn't have one */
      if (!('id' in obj)) {
        obj.id = PLDrawingApi.generateID();
      }

      /* Set the Fabric object's values from what we loaded */
      obj.set({
        left: options.left,
        top: options.top,
        angle: options.angle,
        originX: 'center',
        originY: 'center',
      });

      /* Disable the scaling and rotate controls */
      obj.setControlsVisibility({
        mt: false,
        mb: false,
        ml: false,
        mr: false,
        bl: false,
        br: false,
        tl: false,
        tr: false
      });

      canvas.add(obj);
      submittedAnswer.registerAnswerObject(options, obj);
    });
  }

  static get_button_tooltip(options) {
    return 'Add ' + options.ohm + 'Ω resistor';
  }

  static get_button_icon() {
    return 'resistor-icon.svg';
  }
}

PLDrawingApi.registerElements('resistor', {
  'pl-resistor': Resistor,
});
